implementation/src/main/java/io/smallrye/config/ConfigValue.java

```diff
@@ -3,14 +3,14 @@
 public class ConfigValue {
     private final String name;
     private final String value;
-    private final String configSourceName;
-    private final int configSourceOrdinal;
+    private final String trace;
+    private final ConfigSourceInfo info;
```

Comment:
I don't know if I agree with this part.  To me it makes more sense to add fields for each type of information: for example an `int lineNumber` which can be -1 if there is no line. This field could be overloaded if useful from other configuration types - for example a database might store the primary key here if there is a numeric primary key and if it makes sense to do so.  The other information being stored in `trace` could just as well be a part of the config source name.

What I'm concerned about is that we're increasing the number of objects that have to be created every time a configuration property is requested.  Ideally we'd be doing the opposite.

Reply:
Because we don't know all the ConfigSources, I thought that maybe we would want to leave that generic. A `lineNumber` property seems very specific to some ConfigSources, so I would prefer to have something more generic.

Reply:
True but the harm in having a `lineNumber` that isn't always used is very low compared to the cost of a generic framework which isn't always used.   OTOH the value in being able to introspect the line number is high even if it isn't always set (for example, an application build system like Quarkus might have a common diagnostic infrastructure including line numbers, and being able to extract line numbers (if present) in this case is very useful to provide consistent reporting).

So while it might seem impure from one perspective, it's quite practical from another.

Reply:
And how using a specific object just to hold ConfigSource info? Or do you think everything should be flatten out?

Reply:
I'm thinking that flat is best in this case.